### DISCUSSION TITLE
How to pass passwords etc credentials to mill test?

### DISCUSSION THREAD

**Discussion by jk-1** [com-lihaoyi/mill#2828] [Q&A]
I have few tests which need certain passwords in order to run. I do not want to store these passwords in any file or hard code them into code.

I ported these tests from .sbt -world and with sbt you can implement code that stops and reads from console. Password reading code is pretty much same as solution [here](https://stackoverflow.com/questions/13146019/reading-password-from-console-in-scala).

Problem is that in mill this does not work but throws null pointer exception. It does not make any difference between Command, Task or Input, all throw null pointer exception.

How can I fix this or what is your preferred solution to pass credentials to test code?

**Comment by jk-1** [comment]
I decided to use solution based on environment variables. Couple of good links about the issue.

[How to pass a password to a child process?](https://unix.stackexchange.com/questions/296297/how-to-pass-a-password-to-a-child-process)

[environment variable accessibility in Linux](https://security.stackexchange.com/questions/14000/environment-variable-accessibility-in-linux/14009#14009)

**Reply by lefou** [reply]
Ensure, to access environments variables via the [`T.env` API](https://mill-build.com/api/latest/mill/api/Ctx.html#env:Map[String,String]).
